Add integration tests

EduardSergeev opened the pull request:
Not sure if you need it but I thought it might be helpful to have some tests to check basic functionality of Purple York and to run them via matrix build to check for regression.

This change contains three simple tests defined in [providers.test.ts](https://github.com/EduardSergeev/purple-yolk/blob/b4ea411922c2efe9638d1ef680d82bb1eee80864/source/test/providers.test.ts#L23-L40). Each test:

- Loads specified Haskell file (1st test parameter);
- Waits for the extension to populate diagnostics with specified number of error/warnings (2nd test parameter);
- Runs assertion code (3rd parameter) passing loaded `Document` and corresponding `Diagnostic[]`.

E.g. a test which is defined as:

```typescript
  test('Error only', () => {
    return runFileTest('Error.hs', [DiagnosticSeverity.Error, 1], async (_, diags) => {
      assert.lengthOf(diags, 1);
      assert.include(diags[0]?.message, 'Not in scope: type constructor or class');
    });
  });
```

loads [Error.hs](https://github.com/EduardSergeev/purple-yolk/blob/b4ea411922c2efe9638d1ef680d82bb1eee80864/data/test/Error.hs), waits until the extension produces 1 error and then runs the code with those two assertions.

Test suite can be run via

```console
npm test
```

or with test coverage report (produces by [c8](https://github.com/bcoe/c8)):

```console
npm run coverage
```

Current code coverage:

<pre>=============================== Coverage summary ===============================
<font color="#FCE94F"><b>Statements   : 64.83% ( 780/1203 )</b></font>
<font color="#FCE94F"><b>Branches     : 51.96% ( 53/102 )</b></font>
<font color="#FCE94F"><b>Functions    : 62.85% ( 22/35 )</b></font>
<font color="#FCE94F"><b>Lines        : 64.83% ( 780/1203 )</b></font>
================================================================================
</pre>

And one [from CI](https://github.com/tfausak/purple-yolk/actions/runs/9624266921/job/26547751186#step:6:122)

This change also adds matrix build which runs all tests on all supported platforms with various GHC versions. Here is an [example of new CI run](https://github.com/tfausak/purple-yolk/actions/runs/9624266921).
It also [uploads test coverage](https://github.com/EduardSergeev/purple-yolk/blob/b4ea411922c2efe9638d1ef680d82bb1eee80864/.github/workflows/workflow.yaml#L133-L139) to https://coveralls.io, e.g. [recent coverage report](https://coveralls.io/builds/68239035).

tfausak commented:
Thanks! This is a definite improvement. It would be nice to have some tests to make sure that basic functionality works, at least. However this is also a lot of extra stuff that I'm not familiar with and might not want to maintain. I'll think about it.

EduardSergeev committed:
Fix compillation errors

Due to `null` vs `undefined` mismatches

EduardSergeev committed:
Add integration tests

Add test infrastrucutre
Add a few simple tests to check basic functionality

EduardSergeev committed:
Add matrix build and coverage reporting

Build and test extension on (matrix build):
- ubuntu-latest
- windows-latest
- macos-latest
with:
- 8.10.7
- 9.0.2
- 9.6.5

tfausak commented:
This is great, but it's just more code/infra than I want to maintain for this repo. Thanks though!
